Rename plugins to mixins

jamen opened the pull request:
- Renames `plugins` to `mixins`
 - Allows mixins to extend mixins
 - Updates documentation

jbucaran commented on docs/core.md:
````diff
@@ -313,6 +313,19 @@ app({
+  mixins: [Logger]
 })
 ```
+Mixins can also extend other mixins:
````
@jamen I know it's hard, very hard, to come up with good examples, but since you're introducing this feature, can you try to come up with a more practical example? 🙏

dodekeract replied:
Wouldn't this cause an infinite recursion?

jamen replied:
@dodekeract Only way I could think of is if you intentionally cause a circular reference in the tree of mixins somewhere.  But I don't think that could happen naturally.

jbucaran commented:
@jamen LGTM. I'll merge this after #214. If you can come up with a better example to explain mixins with mixins, that would be a plus. 👍 👍 👍

jbucaran commented:
@jamen With #214 out of the way, let's rebase this, improve the example and merge.

One more thing, the syntax in the docs to indicate array types changed from:

`something[]`

`Array<something>`

dodekeract commented on src/app.js:
```diff
@@ -6,17 +6,21 @@ export default function(app) {
-    if (plugin.state != null) {
-      state = merge(state, plugin.state)
+    if (mixin.mixins != null && mixin !== app) {
+      mixins = mixins.concat(mixin.mixins)
```
That's an awful lot of `mixins`. 😉

jbucaran commented:
@dodekeract You'll recall https://github.com/hyperapp/hyperapp/issues/194 and https://github.com/hyperapp/hyperapp/issues/189.

I've been pretty much swaying towards mixins since the discussions begun. I think @zaceno's argument is sound too.

There's no need to rename plugins to mixins, but now it's the chance to do it, since 1.0.0 is just around the corner.

jamen commented:
Rebased with master and updated the example.  It was hard for me to think of one that fit in a good amount of space, so hopefully this is good enough.  It just relies on the previous `Logger` example.

jamen commented:
I think using a mixin on a mixin is better described as "compose" instead of "extend".  So I changed the description.

lukejacksonn commented on docs/api.md:
```diff
@@ -76,21 +76,22 @@ Type: ([state](#state), [actions](#actions), [data](#update-data), [emit](#emit)
 Fired before the state is updated.
 
-#### render
+#### rendero
```
rendero 😅

jamen replied:
Hah, whoops.  Fixed.

jamen commented on docs/core.md:
```diff
@@ -326,15 +326,15 @@ const Counter = () => ({
-    up: state => { count: state.count + 1 },
-    down: state => { count: state.count + 1 }
+    up: state => ({ count: state.count + 1 }),
+    down: state => ({ count: state.count + 1 })
```
Almost forgot these :sweat_smile:

jamen commented:
Just ran the example through the prettier too.  Was little errors.

jbucaran commented:
@jamen Awesome! Thanks.

jbucaran commented:
@jamen I invited you to join the org 🎉.

BTW I still want to make some tweaks, so I'll publish over the weekend.

jamen commented:
@jbucaran Awesome, thanks! :smile_cat:

Yeah feel free.  Happy to wait for a good v1 release.

terkelg commented:
Nice work @jamen! I really like this change.
